remill/BC/DeadStoreEliminator.h

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2017 Trail of Bits, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef REMILL_BC_DSELIM_H_
+#define REMILL_BC_DSELIM_H_
+
+#include <llvm/IR/InstVisitor.h>
+
+namespace llvm {
+class Type;
+class StructType;
+class Module;
+class DataLayout;
+class Value;
+class AllocaInst;
+class LoadInst;
+class StoreInst;
+class GetElementPtrInst;
+class GetPtrToIntInst;
+class GetIntToPtrInst;
+class BitCastInst;
+//template<typename SubClass> class InstVisitor;
+}  // namespace llvm
+
+namespace remill {
+
+class StateSlot {
+public:
+  StateSlot(uint64_t i, uint64_t offset, uint64_t size);
+  // slot index
+  uint64_t i;
+  // Inclusive beginning byte offset
+  uint64_t offset;
+  // Size of the slot
+  uint64_t size;
+};
+
+class StateVisitor {
+  public:
+    std::vector<StateSlot> slots;
+    // the current index in the state structure
+    uint64_t idx;
+    // the current offset in the state structure
+    uint64_t offset;
+
+    StateVisitor(llvm::DataLayout *dl_);
+
+  private:
+    // the LLVM datalayout used for calculating type allocation size
+    llvm::DataLayout *dl;
+
+  public:
+    // visit a type and record it (and any children) in the slots vector
+    virtual void Visit(llvm::Type *ty);
+};
+
+std::vector<StateSlot> StateSlots(llvm::Module *module);
+
+typedef std::unordered_map<llvm::Value *, uint64_t> ValueToOffset;
+typedef std::unordered_map<llvm::Instruction *, uint64_t> InstToOffset;
+typedef std::unordered_map<llvm::MDNode *, uint64_t> ScopeMap;
+
+enum class VisitResult;
+
+enum class OpType;
+
+bool GetUnsignedOffset(int64_t v1, int64_t v2, OpType op, int64_t max, uint64_t *result);
```

Comment:
We don't need this or `OpType` in the header file because they won't be used outside of your file, so we can make this function `static`.